### PR TITLE
feat(rds-signer): add RDS Signer (#1823)

### PR DESCRIPTION
**Pull request by jgoeglein**
### Issue
#1823 

### Description

Adds RDS Signer for generating a password that can be used for IAM authentication. Previously, this functionality [existed in v2](https://github.com/aws/aws-sdk-js/blob/master/lib/rds/signer.js). 

### Testing

1. Unit tests
2. Built and published tgz locally and deployed project using new module via CDK to verify functionality. [Example manual implementation that was tested is located in this branch](https://github.com/jgoeglein/serverless-patterns/tree/aws-sdk-lib-rds/apigw-http-api-lambda-rds-proxy-cdk/src)

### Additional context

n/a

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/lib-rds/src/Signer.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { Sha256 } from "@aws-crypto/sha256-js";
+import { defaultProvider } from "@aws-sdk/credential-provider-node";
```
Comment (AllanZhengYP) [inline lib/lib-rds/src/Signer.ts:2]: This makes the package requiring all the Node.js specific package, making it not compatible with mobile and browser runtime. Also the `Sha256` is default to pure JS implementation. One thing you can do add a runtime-specific replacement for it. You can look into how it's done in clients:
* [Specify browser and react-native entry in package.json](https://github.com/aws/aws-sdk-js-v3/blob/b0e1422630bbdc6842f7b91629d1ff66693e0fde/clients/client-accessanalyzer/package.json#L84-L89)
* [Throwing error in browsers if credentials are missing](https://github.com/aws/aws-sdk-js-v3/blob/b0e1422630bbdc6842f7b91629d1ff66693e0fde/clients/client-accessanalyzer/src/runtimeConfig.browser.ts#L29)
* [Use `@aws-crypto/sha256-browser` in browser and RN runtimes](https://github.com/aws/aws-sdk-js-v3/blob/b0e1422630bbdc6842f7b91629d1ff66693e0fde/clients/client-accessanalyzer/src/runtimeConfig.browser.ts#L37)
* [Use `@aws-sdk/hash-node` in Node.js runtime for SHA256 util](https://github.com/aws/aws-sdk-js-v3/blob/b0e1422630bbdc6842f7b91629d1ff66693e0fde/clients/client-accessanalyzer/src/runtimeConfig.ts#L46)

File: lib/lib-rds/src/Signer.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      headers: {
+        host: `${this.hostname}:${this.port}`,
+      },
+    });
+
+    const presigned = await signer.presign(request, {
+      expiresIn: 900,
+    });
+
+    return formatUrl(presigned).replace(`${this.protocol}//`, "");
```
Comment (AllanZhengYP) [inline lib/lib-rds/src/Signer.ts:59]: This seems to be confusing. Why is the protocol removed from the full URL?

File: lib/lib-rds/README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# @aws-sdk/lib-rds
+
+[![NPM version](https://img.shields.io/npm/v/@aws-sdk/lib-rds/latest.svg)](https://www.npmjs.com/package/@aws-sdk/lib-rds)
+[![NPM downloads](https://img.shields.io/npm/dm/@aws-sdk/lib-rds.svg)](https://www.npmjs.com/package/@aws-sdk/lib-rds)
+
+## Overview
```
Comment (AllanZhengYP) [inline lib/lib-rds/README.md:6]: Since this package is expected to be consumed directly by the users, a README describing the interface is appreciated.

**Review by AllanZhengYP — COMMENTED** [review]
This very solid start! The interface is mostly complatible with the v2 SDK, which is good. I only have a few suggestions/questions.

File: lib/lib-rds/src/Signer.ts (hunk 2 above)
Comment (jgoeglein) [reply lib/lib-rds/src/Signer.ts:59]: This is a quirk of how RDS implements IAM authentication. [See the botocore notes for their description of this issue here](https://github.com/aws/lumberyard/blob/414466d63520443fafd8c27c4735fee099914384/dev/Tools/AWSPythonSDK/1.5.8/botocore/signers.py#L435).

I will add additional comments to clarify in this repo as well.

Comment (jgoeglein) [reply lib/lib-rds/src/Signer.ts:59]: Additional example from the CLI (note the Output without the leading protocol): https://docs.aws.amazon.com/cli/latest/reference/rds/generate-db-auth-token.html

File: lib/lib-rds/package.json [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    "node": ">= 10.0.0"
+  },
+  "author": {
+    "name": "AWS SDK for JavaScript Team",
+    "url": "https://aws.amazon.com/javascript/"
+  },
+  "license": "Apache-2.0",
+  "dependencies": {
+    "@aws-sdk/signature-v4": "3.40.0",
+    "@aws-sdk/protocol-http": "3.40.0",
```
Comment (AllanZhengYP) [inline lib/lib-rds/package.json:26]: Some dependencies gone missing here: like `@aws-sdk/credential-provider-node`, `@aws-crypto/sha256-js`. This is causing the CI build to fail.

**Comment by AllanZhengYP** [comment]
@jgoeglein Can you move this package to the `packages` folder? The `libs` folder are supposed for abstructions beyond the 1:1 mapping of service APIs. The package doesn't qualify it.

**Review by AllanZhengYP — COMMENTED** [review]
Sorry for waiting so long before I reviewing it again! The PR is in a good shape now. I only have 2 very minor comments. Do you plan to take them? or I can push some commits to this branch.

**Comment by jgoeglein** [comment]
> Sorry for waiting so long before I reviewing it again! The PR is in a good shape now. I only have 2 very minor comments. Do you plan to take them? or I can push some commits to this branch.
> 
> /cc @jgoeglein

@AllanZhengYP  Added you as a collaborator so you can push those changes - it might take me a few days to get back around to this, but happy to do so if needed.

**Comment by AllanZhengYP** [comment]
Okay, I will take it up. For easier interface, I plan to make this signer populating the default credential and region automatically. I post a separate PR https://github.com/aws/aws-sdk-js-v3/pull/3588 to add an easier interface of Node.js default credential provider. I will update thie PR after that one merged.

**Comment by AllanZhengYP** [comment]
According to the latest offline discussion, we decided to standardize all the naming scheme to the signers packages like `s3-request-presigner`, `polly-request-presigner`, and `cloudfront-request-presigner`, because some of them doesn't actually presign a request.

For this package, we decided to keep the name as `rds-signer`, and keep the operation name as `getAuthToken()`.

**Comment by bfaulk96** [comment]
Really looking forward to this addition – Not a fan of importing the entire `aws-sdk` simply to grab an IAM RDS password 😅 

Thanks for all your hard work!

**Review by kuhe — APPROVED** [review]

**Pull request merged**